File: AmazonCareerDay/EntryLevelCodingSD1.py
#EntryLevelCodingSD1.py
#Reference to the interview: https://www.youtube.com/watch?v=595NTKwQbwY
'''
Fist we define the problem:
Amazon is building some stuffs
'''
class StrringChecker():
	def __init__(self,rev,rep,query):
		self.rev = rev
		self.rep = rep
		self.query = query

	def ThreeKeySug(self):
		outputs = []
		lower_rep = sorted([word.lower() for word in self.rep])

		for i in range(1,len(self.query)):
			subQuery = self.query[0:i+1].lower()
			matchedWords = []
			for word in lower_rep:
				if word.startswith(subQuery):
					matchedWords.append(word)
				if len(matchedWords)==3:
					break
				if len(matchedWords) > 0:
					outputs.append(matchedWords)
		return outputs
if __name__ == "__main__":
	#rev: numreviews: 5
	#rep: repository: ["mobile","mouse","moneypot","monitor","mousepad"]
	#query: customerQuery: "mouse"

	#Example 2:
	#rev: 5
	#["code","codePhone","coddle","coddles","codes"]
	#query: coddle
	rev = 5
	rep = ["code","codePhone","coddles","codes"]
	query = "coddle"
	current_string = StrringChecker(rev,rep,query)

	#print (current_string.rev,current_string.rep,current_string.query)
	print (len(current_string.ThreeKeySug()))
	print ('The output for the match is the following:',current_string.ThreeKeySug())

File: AmazonCareerDay/EntryLevelCodingSD1.cpp
//EntryLevelCodingSD1.cpp

/*
Ref: https://www.youtube.com/watch?v=595NTKwQbwY
*/

#include <iostream>
#include <vector>
#include <experimental/algorithm>
#include <string>
#include <boost/algorithm/string.h>

using namespace std;

class StringChecker
{
public:
	//Constructor here
	StringChecker(vector <string> rev, vector <string> rep, vector <string> string query)
	{
		this->rev = rev;
		this->rep = rep ;
		this->query = query;
	}

	//Function to set all characters to lower alphabets
	vector <string> ThreeKeySug(vector <string> rev, vector <string> rep, vector <string> query)
	{
		vector <string> outputs;
		//sort the vector
		vector <string> lower_rep = sort(rep.begin(),rep.end());
		//put all characters to lower case
		std::for_each(lower_rep.begin(),lower_rep.end(),[] (char & c))
		{
			c = ::tolower(c);
		};


		for (int i=1;i<query.size();i++)
		{
			vector <string> subQuery = query[0:i+1];
			std::for_each(subQuery.begin(), subQuery.end(),[] (char & c))
			{
				c = ::tolower(c);

			};
			vector <string> matched_words;
			for (auto word:lower_rep)
			{
				if (word.rfind(subQuery,0)==0)
				{
					matched_words.push_back(word);
				}
				//bool b = boost::algorithm::contains(word,lower_rep);
				if (matched_words.size()==3)
				{
					break;
				}
				if (matched_words.size>0)
				{
					outputs.push_back(matched_words);

				}
			}

		}
		return outputs;
	}
}

int main()
{
	//instantiate the class here

	StringChecker current_string;
	cout << "The output for the match is the following:" << current_string.
	return 0;
}

//TODO: to check and debug the code to see how it works
